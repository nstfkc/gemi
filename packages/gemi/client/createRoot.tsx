import type { ComponentType } from "react";
import { ClientRouter } from "./ClientRouter";
import { ServerDataProvider } from "./ServerDataProvider";

export function createRoot(
  RootLayout: ComponentType<{ children: React.ReactNode; locale: string }>,
) {
  return (props: any) => (
    <ServerDataProvider value={props.data}>
      <ClientRouter
        RootLayout={RootLayout}
        viewImportMap={props.viewImportMap}
      />
    </ServerDataProvider>
  );
}
