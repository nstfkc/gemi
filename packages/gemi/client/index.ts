export { useQuery } from "./useQuery";
export { useMutation } from "./useMutation";
export {
  Form,
  FormError,
  useMutationStatus,
  ValidationErrors,
  FormField,
} from "./Mutation";
export { useLocation, useParams } from "./ClientRouterContext";
export { useSearchParams } from "./useSearchParams";
export { useRoute } from "./useRoute";
export { useIsNavigationPending } from "./useIsNavigationPending";
export { useNavigationProgress } from "./useNavigationProgress";
export { useNavigate } from "./useNavigate";
export { Link } from "./Link";
export { init } from "./init";
export { createRoot } from "./createRoot";

export type { RPC, ViewRPC, I18nDictionary } from "./rpc";
export { type ViewProps, type LayoutProps } from "./types";

export { Image } from "./Image";

export { useForgotPassword } from "./auth/useForgotPassword";
export { useSignIn } from "./auth/useSignIn";
export { useSignUp } from "./auth/useSignUp";
export { useSignOut } from "./auth/useSignOut";
export { useResetPassword } from "./auth/useResetPassword";
export { useUser } from "./auth/useUser";

export { useScopedTranslator, useTranslator } from "./i18n/useScopedTranslator";
export { useLocale } from "./i18n/useLocale";
