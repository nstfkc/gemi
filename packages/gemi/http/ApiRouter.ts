import { isConstructor } from "../internal/isConstructor";
import type { KeyAndValue, KeyAndValueToObject } from "../internal/type-utils";
import {
  Controller,
  ResourceController,
  type ControllerMethods,
} from "./Controller";
import { HttpRequest } from "./HttpRequest";
import type { MiddlewareReturnType } from "./Router";

type HttpMethod = "GET" | "POST" | "PUT" | "PATCH" | "DELETE";

export type ApiRouterHandler<Input, Output, Params> = (
  req: HttpRequest<Input, Params>,
) => Output;

type CallbackHandler<Input, Output, Params> = (
  req: HttpRequest<Input, Params>,
) => Promise<Output> | Output;

type ParseRouteHandler<
  T extends new () => Controller,
  K extends ControllerMethods<T>,
  M extends HttpMethod,
> = InstanceType<T>[K] extends (
  req: HttpRequest<infer Input, infer Params>,
) => infer Output
  ? RouteHandler<M, Input, Output, Params>
  : never;

function isController(
  candidate: CallbackHandler<any, any, any> | (new () => Controller),
): candidate is new () => Controller {
  return isConstructor(candidate);
}

export class RouteHandler<M extends HttpMethod, Input, Output, Params> {
  __internal_brand = "RouteHandler";
  middlewares: string[] = [];

  constructor(
    public method: M,
    private handler:
      | CallbackHandler<Input, Output, Params>
      | (new () => Controller),
    private methodName?: any,
  ) {
    this.handler = handler;
    this.methodName = methodName;
    this.method = method;
  }

  __internal_use() {
    return {
      run: this.run.bind(this),
      middlewares: this.middlewares,
    };
  }

  run(_req: HttpRequest<Input, Params>) {
    if (isController(this.handler)) {
      const controller = new this.handler();
      const handler = controller[this.methodName].bind(controller);
      return handler();
    } else {
      // @ts-ignore
      return this.handler();
    }
  }

  middleware(middlewareList: string[]) {
    this.middlewares = middlewareList;
    return this;
  }
}

export class FileHandler {
  constructor(...args: ConstructorParameters<typeof RouteHandler>) {
    return new RouteHandler(...args) as any;
  }
}

export type RouteHandlers =
  | {
      create: RouteHandler<"POST", any, any, any>;
      list: RouteHandler<"GET", any, any, any>;
    }
  | {
      update: RouteHandler<"PUT", any, any, any>;
      show: RouteHandler<"GET", any, any, any>;
      delete: RouteHandler<"DELETE", any, any, any>;
    };

type IdKey<T> = `/:${T & string}`;

export type ApiRoutes = Record<
  string,
  | RouteHandler<any, any, any, any>
  | FileHandler
  | RouteHandlers
  | typeof ApiRouter
  | ((resourceId: string) => ResourceRoutes<any, string>)
>;

export type ResourceRoutes<
  T extends new () => ResourceController,
  U extends PropertyKey,
> = Record<
  "/",
  {
    list: ParseRouteHandler<T, TestControllerMethod<T, "list">, "GET">;
    create: ParseRouteHandler<T, TestControllerMethod<T, "create">, "POST">;
  }
> &
  Record<
    IdKey<U>,
    {
      show: ParseRouteHandler<T, TestControllerMethod<T, "show">, "GET">;
      update: ParseRouteHandler<T, TestControllerMethod<T, "update">, "PUT">;
      delete: ParseRouteHandler<T, TestControllerMethod<T, "delete">, "DELETE">;
    }
  >;

export class ApiRouter {
  static __brand = "ApiRouter";
  public routes: ApiRoutes = {};
  public middlewares: string[] = [];
  public middleware(_req: HttpRequest<any, any>): MiddlewareReturnType {}

  public get<Input, Output, Params>(
    handler: CallbackHandler<Input, Output, Params>,
  ): RouteHandler<"GET", Input, Output, Params>;
  public get<T extends new () => Controller, K extends ControllerMethods<T>>(
    handler: T,
    methodName: K,
  ): ParseRouteHandler<T, K, "GET">;
  public get<
    T extends CallbackHandler<any, any, any> | (new () => Controller),
    K extends ControllerMethods<any>,
  >(handler: T, methodName?: K) {
    return new RouteHandler("GET", handler, methodName);
  }

  public post<Input, Output, Params>(
    handler: CallbackHandler<Input, Output, Params>,
  ): RouteHandler<"POST", Input, Output, Params>;
  public post<T extends new () => Controller, K extends ControllerMethods<T>>(
    handler: T,
    methodName: K,
  ): ParseRouteHandler<T, K, "POST">;
  public post<
    T extends CallbackHandler<any, any, any> | (new () => Controller),
    K extends ControllerMethods<any>,
  >(handler: T, methodName?: K) {
    return new RouteHandler("POST", handler, methodName);
  }

  public put<Input, Output, Params>(
    handler: CallbackHandler<Input, Output, Params>,
  ): RouteHandler<"PUT", Input, Output, Params>;
  public put<T extends new () => Controller, K extends ControllerMethods<T>>(
    handler: T,
    methodName: K,
  ): ParseRouteHandler<T, K, "PUT">;
  public put<
    T extends CallbackHandler<any, any, any> | (new () => Controller),
    K extends ControllerMethods<any>,
  >(handler: T, methodName?: K) {
    return new RouteHandler("PUT", handler, methodName);
  }

  public patch<Input, Output, Params>(
    handler: CallbackHandler<Input, Output, Params>,
  ): RouteHandler<"PATCH", Input, Output, Params>;
  public patch<T extends new () => Controller, K extends ControllerMethods<T>>(
    handler: T,
    methodName: K,
  ): ParseRouteHandler<T, K, "PATCH">;
  public patch<
    T extends CallbackHandler<any, any, any> | (new () => Controller),
    K extends ControllerMethods<any>,
  >(handler: T, methodName?: K) {
    return new RouteHandler("PATCH", handler, methodName);
  }

  public delete<Input, Output, Params>(
    handler: CallbackHandler<Input, Output, Params>,
  ): RouteHandler<"DELETE", Input, Output, Params>;
  public delete<T extends new () => Controller, K extends ControllerMethods<T>>(
    handler: T,
    methodName: K,
  ): ParseRouteHandler<T, K, "DELETE">;
  public delete<
    T extends CallbackHandler<any, any, any> | (new () => Controller),
    K extends ControllerMethods<any>,
  >(handler: T, methodName?: K) {
    return new RouteHandler("DELETE", handler, methodName);
  }

  public resource<T extends new () => ResourceController>(Controller: T) {
    return <U extends PropertyKey>(resourceId: U) => {
      return {
        "/": {
          list: new RouteHandler("GET", Controller, "list"),
          create: new RouteHandler("POST", Controller, "create"),
        },
        [`/:${String(resourceId)}`]: {
          show: new RouteHandler("GET", Controller, "show"),
          update: new RouteHandler("PUT", Controller, "update"),
          delete: new RouteHandler("DELETE", Controller, "delete"),
        },
      } as ResourceRoutes<T, U>;
    };
  }

  public file<Input, Output, Params>(
    handler: CallbackHandler<Input, Output, Params>,
  ): FileHandler;
  public file<T extends new () => Controller, K extends ControllerMethods<T>>(
    handler: T,
    methodName: K,
  ): FileHandler;
  public file<
    T extends CallbackHandler<any, any, any> | (new () => Controller),
    K extends ControllerMethods<any>,
  >(handler: T, methodName?: K) {
    return new FileHandler("GET", handler, methodName);
  }
}

type TestControllerMethod<T extends new () => Controller, K extends string> =
  K extends ControllerMethods<T> ? K : never;

type RouteHandlerParser<T, Prefix extends string = ""> =
  T extends RouteHandler<infer Method, infer Input, infer Output, infer Params>
    ? KeyAndValue<
        `${Method & string}:${Prefix & string}`,
        ApiRouterHandler<Input, Output, Params>
      >
    : never;

type RouteHandlersParser<
  T,
  Prefix extends string = "",
> = T extends RouteHandlers
  ? {
      [K in keyof T]: T[K] extends RouteHandler<
        infer Method,
        infer Input,
        infer Output,
        infer Params
      >
        ? KeyAndValue<
            `${Method & string}:${Prefix & string}`,
            ApiRouterHandler<Input, Output, Params>
          >
        : never;
    }[keyof T]
  : never;

type RouterInstanceParser<
  T extends new () => ApiRouter,
  Prefix extends string,
> = T extends new () => ApiRouter
  ? RouteParser<InstanceType<T>["routes"], `${Prefix & string}`>
  : never;

type ParsePrefixAndKey<
  P extends PropertyKey,
  K extends PropertyKey,
  U = `${P & string}${K & string}`,
> = U extends "//"
  ? "/"
  : U extends `${infer T1}//${infer T2}`
    ? `${T1}/${T2}`
    : U extends `${infer T1}/${infer T2}/`
      ? `${T1}/${T2}`
      : U;

type LastUrlSegment<T extends string> =
  T extends `${infer _}/${infer LastSegment}` ? LastUrlSegment<LastSegment> : T;

type ToCamelCase<T extends string> = T extends `${infer F}-${infer L}`
  ? `${F}${Capitalize<ToCamelCase<L>>}`
  : T;

type ResourceRoutesParser<
  T extends ResourceRoutes<new () => ResourceController, PropertyKey>,
  U extends PropertyKey,
  K extends keyof T = keyof T,
> = T[K] extends RouteHandlers
  ? K extends "/"
    ? RouteHandlersParser<T[K], `${U & string}`>
    : RouteHandlersParser<
        T[K],
        `${U & string}/:${ToCamelCase<LastUrlSegment<U & string>>}Id`
      >
  : KeyAndValue<"YY", any>;

type RouteParser<
  T extends ApiRoutes,
  Prefix extends PropertyKey = "",
  K extends keyof T = keyof T,
> = K extends any
  ? T[K] extends (
      res: string,
    ) => ResourceRoutes<new () => ResourceController, PropertyKey>
    ? ResourceRoutesParser<ReturnType<T[K]>, ParsePrefixAndKey<Prefix, K>>
    : T[K] extends RouteHandler<any, any, any, any>
      ? RouteHandlerParser<T[K], ParsePrefixAndKey<Prefix, K>>
      : T[K] extends new () => ApiRouter
        ? RouterInstanceParser<T[K], ParsePrefixAndKey<Prefix, K>>
        : T[K] extends RouteHandlers
          ? RouteHandlersParser<
              T[K],
              `${Prefix & string}${K extends "/" ? "" : K & string}`
            >
          : never
  : never;

export type CreateRPC<
  T extends ApiRouter,
  Prefix extends PropertyKey = "",
> = KeyAndValueToObject<RouteParser<T["routes"], Prefix>>;
