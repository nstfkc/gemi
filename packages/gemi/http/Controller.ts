import type { HttpRequest } from "./HttpRequest";

export class Controller {
  requests: Record<string, typeof HttpRequest<any, any>> = {};

  static kind = "controller" as const;

  constructor() {}
}

export type ControllerMethods<T extends new () => Controller> = {
  [K in keyof InstanceType<T>]: InstanceType<T>[K] extends Function ? K : never;
}[keyof InstanceType<T>];
