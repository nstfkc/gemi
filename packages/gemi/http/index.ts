export { Controller, ResourceController } from "./Controller";
export { ApiRouter, type CreateRPC } from "./ApiRouter";
export { ViewRouter } from "./ViewRouter";
export { ValidationError, AuthenticationError } from "./Router";
export { HttpRequest } from "./HttpRequest";
export { Middleware } from "./Middleware";
export { getCookies } from "./getCookies";
export { RequestBreakerError } from "./Error";
