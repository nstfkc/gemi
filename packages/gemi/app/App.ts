import type { ApiRouteExec, ApiRouter } from "../http/ApiRouter";
import { ViewRouter, type ViewRouteExec } from "../http/ViewRouter";
// @ts-ignore
import { URLPattern } from "urlpattern-polyfill";
import { generateETag } from "../server/generateEtag";
import { v4 } from "uuid";
import type { ComponentTree } from "../client/types";
import { type RouterMiddleware } from "../http/Router";
import type { UnwrapPromise } from "../utils/type";
import { GEMI_REQUEST_BREAKER_ERROR } from "../http/Error";
import type { Plugin } from "./Plugin";
import type { Middleware } from "../http/Middleware";
import { requestContext } from "../http/requestContext";
import type { ServerWebSocket } from "bun";
import { flattenComponentTree } from "../client/helpers/flattenComponentTree";

import { createComponentTree } from "./createComponentTree";
import { createFlatViewRoutes } from "./createFlatViewRoutes";
import { createRouteManifest } from "./createRouteManifest";
import { createFlatApiRoutes } from "./createFlatApiRoutes";

// @ts-ignore
import { renderToReadableStream } from "react-dom/server.browser";
import { ComponentType, createElement, Fragment } from "react";

import { isConstructor } from "../internal/isConstructor";
import { HttpRequest } from "../http";

interface RenderParams {
  styles: string[];
  manifest: Record<string, any>;
  serverManifest: Record<string, any>;
  bootstrapModules?: string[];
}

interface AppParams {
  viewRouter: new () => ViewRouter;
  apiRouter: new () => ApiRouter;
  plugins?: (new () => Plugin)[];
  middlewareAliases?: Record<string, new () => Middleware>;
  root: ComponentType;
}

export class App {
  private renderParams: RenderParams = {
    styles: [],
    manifest: {},
    serverManifest: {},
    bootstrapModules: [],
  };
  private flatViewRoutes: Record<
    string,
    { exec: ViewRouteExec[]; middleware: any[] }
  > = {};
  private flatApiRoutes: Record<
    string,
    Record<
      string,
      {
        exec: ApiRouteExec;
        middleware: (string | (new () => Middleware) | RouterMiddleware)[];
      }
    >
  > = {};
  private routeManifest: Record<string, string[]> = {};
  public name = "APP";
  private appId: string;
  private componentTree: ComponentTree;
  public middlewareAliases: Record<string, new () => Middleware> = {};
  public devVersion = 0;
  private params: AppParams;
  private apiRouter: new () => ApiRouter;
  private viewRouter: new () => ViewRouter;
  private Root: ComponentType;

  constructor(params: AppParams) {
    this.params = params;
    this.apiRouter = params.apiRouter;
    this.viewRouter = params.viewRouter;
    this.Root = params.root;

    this.prepare();
    this.appId = generateETag(Date.now());
  }

  private prepare() {
    const params = this.params;
    this.middlewareAliases = params.middlewareAliases ?? {};

    let viewRouters = {
      "/": this.viewRouter,
    };
    let apiRouters = {
      "/": this.apiRouter,
    };
    for (const Plugin of params.plugins ?? []) {
      const plugin = new Plugin();
      if (plugin.viewRouter) {
        viewRouters = {
          ...viewRouters,
          [plugin.viewRoutesBasePath]: plugin.viewRouter,
        };
      }
      if (plugin.apiRouter) {
        apiRouters = {
          ...apiRouters,
          [plugin.apiRoutesBasePath]: plugin.apiRouter,
        };
      }
    }
    this.flatViewRoutes = createFlatViewRoutes(viewRouters);
    this.componentTree = createComponentTree(viewRouters);
    this.routeManifest = createRouteManifest(viewRouters);
    // Handle view routes

    // Handle api routes
    this.flatApiRoutes = createFlatApiRoutes(apiRouters);
  }

  public getComponentTree() {
    return this.componentTree;
  }

  public setRenderParams(params: RenderParams) {
    this.renderParams = params;
  }

  private async resolvePageData(req: Request) {
    const url = new URL(req.url);
    let handlers: ViewRouteExec[] = [];
    let middlewares: (RouterMiddleware | string)[] = [];
    let currentPathName = "";
    let params: Record<string, any> = {};
    const sortedEntries = Object.entries(this.flatViewRoutes).sort(
      ([pathA], [pathB]) => pathB.length - pathA.length,
    );

    for (const [pathname, handler] of sortedEntries) {
      const pattern = new URLPattern({ pathname });
      if (pattern.test({ pathname: url.pathname })) {
        currentPathName = pathname;
        params = pattern.exec({ pathname: url.pathname })?.pathname.groups!;
        handlers = handler.exec;
        middlewares = handler.middleware;
      }
    }

    const reqWithMiddlewares = this.runMiddleware(middlewares);

    const reqCtx = new Map();

    const data = await requestContext.run(reqCtx, async () => {
      await reqWithMiddlewares(new HttpRequest(req), reqCtx);
      return await Promise.all(handlers.map((fn) => fn(req, params, this)));
    });

    return {
      data,
      currentPathName,
      params,
      user: reqCtx.get("user") ?? null,
    };
  }

  private runMiddleware(
    middleware: (string | RouterMiddleware | (new () => Middleware))[],
  ) {
    return middleware
      .map((aliasOrTest) => {
        if (typeof aliasOrTest === "string") {
          const alias = aliasOrTest;
          if (this.middlewareAliases?.[alias]) {
            const middleware = new this.middlewareAliases[alias]();
            return middleware.run;
          }
        } else {
          if (isConstructor(aliasOrTest)) {
            // TODO: fix type
            // @ts-ignore
            const middleware = new aliasOrTest();
            return middleware.run;
          }
          return aliasOrTest;
        }
      })
      .filter(Boolean)
      .reduce(
        (acc: any, middleware: any) => {
          return async (req: HttpRequest, ctx: any) => {
            return {
              ...(await acc(req, ctx)),
              ...(await middleware(req, ctx)),
            };
          };
        },
        (_req: HttpRequest) => Promise.resolve({}),
      );
  }

  async handleApiRequest(req: Request) {
    const url = new URL(req.url);
    const apiPath = url.pathname.replace("/api", "");
    for (const [path, handler] of Object.entries(this.flatApiRoutes)) {
      const pattern = new URLPattern({ pathname: path });
      if (pattern.test({ pathname: apiPath })) {
        const params = pattern.exec({ pathname: apiPath })?.pathname.groups!;
        if (!handler[req.method.toLowerCase()]) {
          return new Response(
            JSON.stringify({ error: { message: "Not found" } }),
          );
        }
        const exec = handler[req.method.toLowerCase()].exec;
        const middlewares = handler[req.method.toLowerCase()].middleware;

        const reqWithMiddlewares = this.runMiddleware(middlewares);

        const reqCtx = new Map();

        return await requestContext.run(reqCtx, async () => {
          const httpRequest = new HttpRequest(req);
          let handler = exec
            ? () => exec(httpRequest, params, this)
            : () => Promise.resolve({});

          try {
            await reqWithMiddlewares(httpRequest, reqCtx);
          } catch (err) {
            if (err.kind === GEMI_REQUEST_BREAKER_ERROR) {
              const { status = 400, data, headers } = err.payload.api;

              return new Response(JSON.stringify(data), {
                status,
                headers: {
                  "Content-Type": "application/json",
                  ...headers,
                },
              });
            } else {
              throw err;
            }
          }

          let data = {};
          try {
            data = await handler();
          } catch (err) {
            if (err.kind === GEMI_REQUEST_BREAKER_ERROR) {
              const { status = 400, data, headers } = err.payload.api;

              return new Response(JSON.stringify(data), {
                status,
                headers: {
                  "Content-Type": "application/json",
                  ...headers,
                },
              });
            } else {
              throw err;
            }
          }

          const result = {
            kind: "api",
            data,
            headers: {
              "Set-Cookie": Object.entries({})
                .map(([name, config]) => {
                  return `${name}=${(config as any).value}; HttpOnly; SameSite=Strict; Path=/; Max-Age=${(config as any).maxAge}`;
                })
                .join(", "),
            },
          };
          return new Response(JSON.stringify(result.data), {
            headers: {
              "Content-Type": "application/json",
              ...result.headers,
            },
          });
        });
      }
    }

    return new Response(JSON.stringify({ error: { message: "Not found" } }), {
      status: 404,
    });
  }

  async handleViewRequest(req: Request) {
    const url = new URL(req.url);

    let pageData: UnwrapPromise<
      ReturnType<typeof this.resolvePageData>
    > | null = null;

    try {
      let handlers: ViewRouteExec[] = [];
      let middlewares: (RouterMiddleware | string)[] = [];
      let currentPathName: null | string = null;
      let params: Record<string, any> = {};
      const sortedEntries = Object.entries(this.flatViewRoutes).sort(
        ([pathA], [pathB]) => pathB.length - pathA.length,
      );

      for (const [pathname, handler] of sortedEntries) {
        const pattern = new URLPattern({ pathname });
        if (pattern.test({ pathname: url.pathname })) {
          currentPathName = pathname;
          params = pattern.exec({ pathname: url.pathname })?.pathname.groups!;
          handlers = handler.exec;
          middlewares = handler.middleware;
        }
      }

      const reqWithMiddlewares = this.runMiddleware(middlewares);

      const reqCtx = new Map();

      const data = await requestContext.run(reqCtx, async () => {
        await reqWithMiddlewares(new HttpRequest(req), reqCtx);
        return await Promise.all(handlers.map((fn) => fn(req, params, this)));
      });

      pageData = {
        data,
        currentPathName,
        params,
        user: reqCtx.get("user") ?? null,
      };
    } catch (err) {
      if (err.kind === GEMI_REQUEST_BREAKER_ERROR) {
        return new Response(null, {
          ...err.payload.view,
        });
      }
    }

    // TODO: handle 404
    if (!pageData) {
      return new Response("<div>404</div>");
    }

    const { data, params, currentPathName, user } = pageData;

    const viewData = data.reduce((acc, data) => {
      return {
        ...acc,
        ...data,
      };
    }, {});

    if (url.searchParams.get("json")) {
      const result = {
        kind: "viewData",
        data: {
          [url.pathname]: viewData,
        },
        meta: [],
        headers: {},
      };

      return new Response(
        JSON.stringify({
          data: result.data,
          head: {},
        }),
        {
          headers: {
            ...result.headers,
            "Content-Type": "application/json",
          },
        },
      );
    }

    let cookieHeaders = {};
    const visitorIdExist = req.headers.get("cookie")?.includes("visitor_id");
    if (!visitorIdExist) {
      cookieHeaders = {
        "Set-Cookie": `visitor_id=${v4()}; HttpOnly; SameSite=Strict; Path=/`,
      };
    }

    const { styles, manifest, serverManifest, bootstrapModules } =
      this.renderParams;

    const viewImportMap = {};
    let appDir: string | null = null;
    const template = (viewName: string, path: string) =>
      `"${viewName}": () => import("${path}")`;
    const templates = [];

    for (const fileName of [
      "404",
      ...flattenComponentTree(this.componentTree),
    ]) {
      if (process.env.NODE_ENV === "test") {
        break;
      }
      if (!manifest) {
        appDir = `${process.env.APP_DIR}`;
        const mod = await import(`${appDir}/views/${fileName}.tsx`);
        viewImportMap[fileName] = mod.default;
        templates.push(template(fileName, `${appDir}/views/${fileName}.tsx`));
      } else {
        const serverFile = serverManifest[`app/views/${fileName}.tsx`];
        const mod = await import(
          `${process.env.DIST_DIR}/server/${serverFile?.file}`
        );
        viewImportMap[fileName] = mod.default;
        const clientFile = manifest[`app/views/${fileName}.tsx`];
        if (clientFile) {
          templates.push(template(fileName, `/${clientFile?.file}`));
        }
      }
    }

    const loaders = `{${templates.join(",")}}`;

    const result = {
      kind: "view",
      data: {
        pageData: {
          [url.pathname]: viewData,
        },
        auth: { user },
        routeManifest: this.routeManifest,
        router: {
          pathname: currentPathName,
          params,
          currentPath: url.pathname,
          is404: !currentPathName ? true : false,
        },
        componentTree: [["404", []], ...this.componentTree],
      },
      head: {},
      headers: {
        "Content-Type": "text/html",
        "Cache-Control": user
          ? "private, no-cache, no-store, max-age=0, must-revalidate"
          : "public, max-age=864000, must-revalidate",
        ETag: this.appId,
        ...cookieHeaders,
      },
      status: !currentPathName ? 404 : 200,
    };

    const stream = await renderToReadableStream(
      createElement(Fragment, {
        children: [
          styles,
          createElement(this.Root as any, {
            data: result.data,
            viewImportMap,
          }),
        ],
      }),
      {
        bootstrapScriptContent: `window.__GEMI_DATA__ = ${JSON.stringify(result.data)}; window.loaders=${loaders}`,
        bootstrapModules,
      },
    );

    return new Response(stream, {
      status: result.status,
      headers: {
        "Content-Type": "text/html",
        ...result.headers,
      },
    });
  }

  async fetch(req: Request): Promise<Response> {
    const url = new URL(req.url);

    if (url.pathname.startsWith("/api")) {
      return await this.handleApiRequest(req);
    } else {
      return await this.handleViewRequest(req);
    }
  }

  private handleWebSocketMessage = (
    ws: ServerWebSocket,
    message: string | Buffer,
  ) => {};
  private handleWebSocketOpen = (ws: ServerWebSocket) => {
    console.log("socket opened");
  };
  private handleWebSocketClose = (
    ws: ServerWebSocket,
    code: number,
    reason: string,
  ) => {
    console.log("socket closed");
  };

  websocket = {
    message: this.handleWebSocketMessage,
    open: this.handleWebSocketOpen,
    close: this.handleWebSocketClose,
  };
}
