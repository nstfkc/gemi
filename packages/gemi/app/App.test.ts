import { describe, test, expect } from "vitest";

import { App } from "./App";
import { ApiRouter } from "../http/ApiRouter";
import { ViewRouter } from "../http/ViewRouter";

import { createRoot } from "../client/createRoot";
import { createElement } from "react";
import { Controller } from "../http/Controller";
import { HttpRequest } from "../http/HttpRequest";
import { Middleware, RequestBreakerError } from "../http";
import { Kernel } from "../kernel";

class ValidationRequest extends HttpRequest<{ name: string; age: number }, {}> {
  schema = {
    name: {
      required: "Name is required",
      string: "Name must be a string",
    },
    age: {
      required: "Age is required",
      number: "Age must be a number",
    },
  };
}

class TestController extends Controller {
  requests = {
    validate: ValidationRequest,
  };

  test() {
    return { message: "test" };
  }
  async foo(req: HttpRequest) {
    const body = await req.input();

    return body.toJSON();
  }

  async validate(req: ValidationRequest) {
    const body = await req.input();

    return body.toJSON();
  }
}

class RootApiRouter extends ApiRouter {
  routes = {
    "/test": this.get(() => {
      return { message: "hi" };
    }),
    "/foo": this.post(TestController, "foo"),
    "/bar": this.get(TestController, "test").middleware(["auth"]),
    "/baz": this.post(TestController, "validate"),
  };
}

class RootViewRouter extends ViewRouter {
  routes = {
    "/": this.view("Home", () => {
      return { message: "Home" };
    }),
    "/about": this.view("About", [TestController, "test"]),
  };
}

export class AuthenticationError extends RequestBreakerError {
  constructor() {
    super("Authentication error");
    this.name = "AuthenticationError";
  }

  payload = {
    api: {
      status: 401,
      data: { error: "Authentication error" },
    },
    view: {
      status: 302,
      headers: {
        "Cache-Control":
          "private, no-cache, no-store, max-age=0, must-revalidate",
        Location: "/auth/sign-in",
      },
    },
  };
}

class AuthMiddleware extends Middleware {
  async run(req: HttpRequest) {
    throw new AuthenticationError();

    return {};
  }
}

const app = new App({
  root: createRoot(() => createElement("div")),
  apiRouter: RootApiRouter,
  viewRouter: RootViewRouter,
  kernel: Kernel,
  middlewareAliases: {
    auth: AuthMiddleware,
  },
});

describe("App fetch()", () => {
  test("view", async () => {
    const res1 = await app.fetch(
      new Request("http://gemi.dev", {
        method: "GET",
      }),
    );

    expect(await res1.text()).toMatchSnapshot();
  });

  test("api", async () => {
    const res = await app.fetch(
      new Request("http://gemi.dev/api/test", {
        method: "GET",
      }),
    );

    expect(await res.json()).toEqual({ message: "hi" });
  });

  test("view controller handler", async () => {
    const request = new Request("http://gemi.dev/about?json=true", {
      method: "GET",
    });
    const res = await app.fetch(request);
    expect(await res.json()).toEqual({
      data: { "/about": { About: { message: "test" } } },
      head: {},
    });
  });

  test("api callback handler", async () => {
    const request = new Request("http://gemi.dev/api/test", {
      method: "GET",
    });
    const res = await app.fetch(request);

    expect(await res.json()).toEqual({ message: "hi" });
  });

  test("api controller handler", async () => {
    const request = new Request("http://gemi.dev/api/foo", {
      method: "POST",
      body: JSON.stringify({ data: 1 }),
      headers: {
        "Content-Type": "application/json",
      },
    });
    const res = await app.fetch(request);
    expect(await res.json()).toEqual({ data: 1 });
  });

  test("api handler with middleware", async () => {
    const request = new Request("http://gemi.dev/api/bar", {
      method: "GET",
      body: JSON.stringify({ data: 1 }),
      headers: {
        "Content-Type": "application/json",
      },
    });
    const res = await app.fetch(request);

    expect(await res.json()).toEqual({ error: "Authentication error" });
    expect(res.status).toBe(401);
  });

  test("api handler with validation", async () => {
    const request = new Request("http://gemi.dev/api/baz", {
      method: "POST",
      body: JSON.stringify({ age: "12" }),
      headers: {
        "Content-Type": "application/json",
      },
    });
    const res = await app.fetch(request);

    expect(res.status).toBe(400);
    expect(await res.json()).toEqual({
      error: {
        kind: "validation_error",
        messages: {
          age: ["Age must be a number"],
          name: ["Name is required"],
        },
      },
    });
  });

  test("404 api handler", async () => {
    const request = new Request("http://gemi.dev/api/404", {
      method: "GET",
    });
    const res = await app.fetch(request);
    expect(res.status).toBe(404);
    expect(await res.json()).toEqual({ error: { message: "Not found" } });
  });

  test("404 view handler", async () => {
    const request = new Request("http://gemi.dev/not-found", {
      method: "GET",
    });
    const res = await app.fetch(request);
    expect(res.status).toBe(404);
  });
});
