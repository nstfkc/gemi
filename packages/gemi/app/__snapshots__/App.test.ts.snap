// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`App fetch() view 1`] = `"<div></div><script>window.__GEMI_DATA__ = {"pageData":{"/":{"Home":{"message":"Home"}}},"auth":{"user":null},"routeManifest":{"/":["Home"],"/about":["About"],"/auth/sign-in":["auth/SignIn"],"/auth/sign-up":["auth/SignUp"],"/auth/reset-password":["auth/ResetPassword"],"/auth/forgot-password":["auth/ForgotPassword"]},"router":{"pathname":"/","params":{},"currentPath":"/","searchParams":"","is404":false},"componentTree":[["404",[]],["Home",[]],["About",[]],["auth/SignIn",[]],["auth/SignUp",[]],["auth/ResetPassword",[]],["auth/ForgotPassword",[]]]}; window.loaders={}</script>"`;
